Mathlib/Probability/Kernel/Disintegration/Basic.lean

```diff
@@ -0,0 +1,459 @@
+      = condKernelUnitBorel (const Unit ρ) := by
+    ext
+    simp only [prodMkLeft_apply, Measure.condKernel_apply]
+  rw [Measure.compProd, h1, h2, compProd_fst_condKernelUnitBorel]
+  simp
+#align probability_theory.measure_eq_comp_prod MeasureTheory.Measure.compProd_fst_condKernel
+
+/-- Auxiliary lemma for `condKernel_apply_of_ne_zero`. -/
+lemma _root_.MeasureTheory.Measure.condKernel_apply_of_ne_zero_of_measurableSet
+    [MeasurableSingletonClass α] {x : α} (hx : ρ.fst {x} ≠ 0) {s : Set Ω} (hs : MeasurableSet s) :
+    ρ.condKernel x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s) := by
+  nth_rewrite 3 [← ρ.compProd_fst_condKernel]
+  rw [Measure.compProd_apply (measurableSet_prod.mpr (Or.inl ⟨measurableSet_singleton x, hs⟩))]
+  classical
+  have : ∀ a, ρ.condKernel a (Prod.mk a ⁻¹' {x} ×ˢ s)
+      = ({x} : Set α).indicator (fun a ↦ ρ.condKernel a s) a := by
+    intro a
+    by_cases hax : a = x
+    · simp only [hax, singleton_prod, mem_singleton_iff, indicator_of_mem]
+      congr with y
+      simp
+    · simp only [singleton_prod, mem_singleton_iff, hax, not_false_eq_true, indicator_of_not_mem]
+      have : Prod.mk a ⁻¹' (Prod.mk x '' s) = ∅ := by
+        ext y
+        simp [Ne.symm hax]
+      simp only [this, measure_empty]
+  simp_rw [this]
+  rw [MeasureTheory.lintegral_indicator _ (measurableSet_singleton x)]
+  simp only [Measure.restrict_singleton, lintegral_smul_measure, lintegral_dirac]
+  rw [← mul_assoc, ENNReal.inv_mul_cancel hx (measure_ne_top ρ.fst _), one_mul]
+
+/-- If the singleton `{x}` has non-zero mass for `ρ.fst`, then for all `s : Set Ω`,
+`ρ.condKernel x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s)` . -/
+lemma _root_.MeasureTheory.Measure.condKernel_apply_of_ne_zero [MeasurableSingletonClass α]
+    {x : α} (hx : ρ.fst {x} ≠ 0) (s : Set Ω) :
+    ρ.condKernel x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s) := by
+  have : ρ.condKernel x s = ((ρ.fst {x})⁻¹ • ρ).comap (fun (y : Ω) ↦ (x, y)) s := by
+    congr 2 with s hs
+    simp [Measure.condKernel_apply_of_ne_zero_of_measurableSet hx hs,
+      (measurableEmbedding_prod_mk_left x).comap_apply]
+  simp [this, (measurableEmbedding_prod_mk_left x).comap_apply, hx]
+
+end Measure
+
+section Countable
+
+variable [Countable α]
+
+/-- Auxiliary definition for `ProbabilityTheory.kernel.condKernel`.
+A conditional kernel for `κ : kernel α (β × Ω)` where `α` is countable and `Ω` is standard Borel. -/
+noncomputable
+def condKernelCountable (κ : kernel α (β × Ω)) [IsFiniteKernel κ] : kernel (α × β) Ω where
+  val p := (κ p.1).condKernel p.2
+  property := by
+    change Measurable ((fun q : β × α ↦ (κ q.2).condKernel q.1) ∘ Prod.swap)
+    refine (measurable_from_prod_countable' (fun a ↦ ?_) ?_).comp measurable_swap
+    · exact kernel.measurable (κ a).condKernel
+    · intro y y' x hy'
+      have : κ y' = κ y := by
+        ext s hs
+        exact mem_of_mem_measurableAtom hy'
+          (kernel.measurable_coe κ hs (measurableSet_singleton (κ y s))) rfl
```

Comment:
@sgouezel That's the part of your explanation I did not quite get before. Now I see it. Thanks a lot!